File: p2/quicksort.pl
#
# Copyright (c) 2005-2016 by Xiaoye Meng.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

sub partition {
    my ($a, $p, $r) = @_;
    my $i = $p - 1;
    for my $j ($p..$r - 1) {
        if ($a->[$j] <= $a->[$r]) {
            ++$i;
            @$a[$i, $j] = @$a[$j, $i];
        }
    }
    @$a[$i + 1, $r] = @$a[$r, $i + 1];
    $i + 1;
}

sub quicksort {
    my ($a, $p, $r) = @_;
    if ($p < $r) {
        my $q = partition($a, $p, $r);
        quicksort($a, $p, $q - 1);
        quicksort($a, $q + 1, $r);
    }
}

sub randomized_partition {
    my ($a, $p, $r) = @_;
    my $i = int(rand($r - $p + 1)) + $p;
    @$a[$r, $i] = @$a[$i, $r];
    partition($a, $p, $r);
}

sub randomized_quicksort {
    my ($a, $p, $r) = @_;
    if ($p < $r) {
        my $q = randomized_partition($a, $p, $r);
        randomized_quicksort($a, $p, $q - 1);
        randomized_quicksort($a, $q + 1, $r);
    }
}

sub hoare_partition {
    my ($a, $p, $r) = @_;
    my ($i, $j, $x) = ($p, $r, $a->[$p]);
    while (1) {
        ++$i while $a->[$i] < $x;
        --$j while $a->[$j] > $x;
        $i < $j ? @$a[$i, $j] = @$a[$j, $i] : return $j;
    }
}

File: p2/quicksort.go
//
// Copyright (c) 2016 by Qiang Meng.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

package p2

import "math/rand"

func Quicksort(a []int, p, r int) {
    if p < r {
        q := partition(a, p, r)
        Quicksort(a, p, q - 1)
        Quicksort(a, q + 1, r)
    }
}

func partition(a []int, p, r int) int {
    i := p - 1
    for j := p; j < r; j++ {
        if a[j] <= a[r] {
            i++
            a[i], a[j] = a[j], a[i]
        }
    }
    a[i + 1], a[r] = a[r], a[i + 1]
    return i + 1
}

func randomizedPartition(a []int, p, r int) int {
    i := rand.Intn(r - p + 1) + p
    a[r], a[i] = a[i], a[r]
    return partition(a, p, r)
}

func RandomizedQuicksort(a []int, p, r int) {
    if p < r {
        q := randomizedPartition(a, p, r)
        RandomizedQuicksort(a, p, q - 1)
        RandomizedQuicksort(a, q + 1, r)
    }
}

func hoarePartition(a []int, p, r int) int {
    i, j, x := p, r, a[p]
    for true {
        for a[i] < x {
            i++
        }
        for a[j] > x {
            j--
        }
        if i < j {
            a[i], a[j] = a[j], a[i]
        } else {
            break
        }
    }
    return j
}

